porch/repository/pkg/git/git.go

```diff
@@ -754,3 +779,50 @@ func (r *gitRepository) pushAndCleanup(ctx context.Context, ph *pushRefSpecBuild
 	}
 	return nil
 }
+
+func (r *gitRepository) loadTasks(ctx context.Context, startCommit *object.Commit) ([]v1alpha1.Task, error) {
```

Comment:
Is the idea that even published packages have tasks? Currently when publishing package, history of the intermediate commits is thrown away. We only maintain it for draft authoring; on approval it is collapsed away.

go-git has some commit-tree walking iterators too that can be helpful, I _think_ even including iterators that avoid specified ancestor paths...

Reply:
I think it's a good question re published packages.  If we think the information is valuable during authoring and approval (which I do), I think we should keep the information.  e.g. Imagine this scenario: package update goes through, deployed instances are updated, one of them breaks, we want to understand why: how was the bad yaml introduced, was it adequately presented in review, how can we prevent this happening again?  We could archive the information if we wanted to squash, but I'm not sure what it buys us really!  Also I think when a deployed instance gets a request to update, it will want to know what the upstream changes are.  My view is that the tasks are an important way to make sense of what has changed, adding useful insight over the raw YAML changes.

I propose adding a TODO re published packages, we can see how it feels in actual usage.

I will take a look at the iterators - it's a great point.

Reply:
TODO for published packages is a good idea. I mostly want to make sure the code is self-consistent (i.e. we don't have some parts of the code assume that the history for published packages exists while it in actuality doesn't)